File: TreiVieti/Screen/blood groups/BloodGroupCell.m
//
//  BloodGroupNeedView.m
//  TreiVieti
//
//  Created by Ifrim Alexandru on 6/22/19.
//  Copyright © 2019 Tudor Turcanu. All rights reserved.
//

#import "BloodGroupCell.h"
#import <TreiVieti-Swift.h>
#import "UIColor+Extensions.h"

@implementation BloodGroupCell
@synthesize record = _record;

-(id)initWithFrame:(CGRect)frame {
    // frame.size.height = 82
    self = [super initWithFrame:frame];
    if (!self) {
        return nil;
    }
    
    self.clipsToBounds = NO;
    self.backgroundColor = [UIColor clearColor];
    self.layer.cornerRadius = 5;
    self.layer.shadowOffset = CGSizeMake(0, 5);
    self.layer.shadowColor = [UIColor colorWithWhite:0 alpha:0.2f].CGColor;
    self.layer.shadowOpacity = 1;
    self.layer.shadowRadius = 5;

    _lblBloodGroup = [[UILabel alloc] initWithFrame:CGRectMake(0, 10, frame.size.width, 29)];;
    _lblBloodGroup.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleBottomMargin;
    _lblBloodGroup.textAlignment = NSTextAlignmentCenter;
    _lblBloodGroup.textColor = [UIColor appTextColor];
    _lblBloodGroup.font = [UIFont fontWithName:@"Poppins-Bold" size:30];
    [self addSubview:_lblBloodGroup];
    
    _lblFillPercentage = [[UILabel alloc] initWithFrame:CGRectMake(0, 53, frame.size.width, 29)];;
    _lblFillPercentage.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleTopMargin;
    _lblFillPercentage.textAlignment = NSTextAlignmentCenter;
    _lblFillPercentage.textColor = [UIColor appTextColor];
    _lblFillPercentage.font = [UIFont fontWithName:@"Poppins-Regular" size:14];
    [self addSubview:_lblFillPercentage];
    
    _fillColor = [UIColor appRedColor];
    
    return self;
}

#pragma mark - draw
-(void)drawRect:(CGRect)rect {
    [super drawRect:rect];
    
    CGRect fillRect = self.bounds;
    fillRect.size.height = roundf(fillRect.size.height * _record.fillPercentage / 100.0f);
    fillRect.origin.y = self.bounds.size.height - fillRect.size.height;
    
    CGContextRef context = UIGraphicsGetCurrentContext();

    // white fill
    CGPathRef fillPath = [UIBezierPath bezierPathWithRoundedRect:self.bounds cornerRadius:5].CGPath;
    CGContextAddPath(context, fillPath);
    CGContextSetFillColorWithColor(context, [UIColor whiteColor].CGColor);
    CGContextFillPath(context);
    
    // red fill
    fillPath = [UIBezierPath bezierPathWithRoundedRect:fillRect
                      
                                           byRoundingCorners:UIRectCornerBottomLeft | UIRectCornerBottomRight
                                                 cornerRadii:CGSizeMake(5, 5)].CGPath;
    CGContextAddPath(context, fillPath);
    CGContextSetFillColorWithColor(context, _fillColor.CGColor);
    CGContextFillPath(context);
}

#pragma mark - properties
-(void)setRecord:(BloodGroup *)record {
    _record = record;
    [self refresh];
}
-(void)refresh {
    _lblBloodGroup.text = _record.groupName;
    _lblFillPercentage.text = [NSString stringWithFormat:@"%d%%", _record.fillPercentage];
    [self setNeedsLayout];
}
@end

File: TreiVieti/Screen/Home/blood groups/BloodGroupCell.swift
//
//  BloodGroupCell.swift
//  TreiVieti
//
//  Created by Ifrim Alexandru on 6/23/19.
//  Copyright © 2019 Tudor Turcanu. All rights reserved.
//

import UIKit

class BloodGroupCell : UICollectionViewCell {
    var record : BloodGroupModel? {
        didSet {
            self.refresh()
        }
    }
    var lblBloodGroup : UILabel?
    var lblFillPercentage : UILabel?
    var fillColor : UIColor = UIColor.appRedColor
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.commonInit()
    }
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.commonInit();
    }
    func commonInit() {
        self.clipsToBounds = false
        self.backgroundColor = UIColor.clear
        self.layer.cornerRadius = 5
        self.layer.shadowOffset = CGSize(width: 0, height: 5)
        self.layer.shadowColor = UIColor(white: 0, alpha: 0.2).cgColor
        self.layer.shadowOpacity = 1
        self.layer.shadowRadius = 5
        
        let labelTitle = UILabel(frame: CGRect(x: 0, y: 10, width: frame.self.width, height: 29))
        labelTitle.autoresizingMask = [.flexibleWidth, .flexibleBottomMargin]
        labelTitle.textAlignment = NSTextAlignment.center;
        labelTitle.textColor = UIColor.appTextColor
        labelTitle.font = UIFont(name: "Poppins-Bold", size: 30)
        self.addSubview(labelTitle)
        self.lblBloodGroup = labelTitle
        
        let labelFill = UILabel(frame: CGRect(x: 0, y: 53, width: frame.size.width, height: 29))
        labelFill.autoresizingMask = [.flexibleWidth, .flexibleTopMargin]
        labelFill.textAlignment = NSTextAlignment.center
        labelFill.textColor = UIColor.appTextColor
        labelFill.font = UIFont(name: "Poppins-Regular", size: 14)
        self.addSubview(labelFill)
        self.lblFillPercentage = labelFill
    }
    
    // MARK: draw
    override func draw(_ rect: CGRect) {
        super.draw(rect)
        
        let context = UIGraphicsGetCurrentContext()!
        
        // white background fill
        var fillPath : CGPath = UIBezierPath(roundedRect: self.bounds, cornerRadius: 5).cgPath
        context.addPath(fillPath)
        context.setFillColor(UIColor.white.cgColor)
        context.fillPath()
        
        // red fill
        var fillRect : CGRect = self.bounds;
        fillRect.size.height = CGFloat(roundf(Float(fillRect.size.height * CGFloat(self.record!.fillPercentage) / 100.0)))
        fillRect.origin.y = self.bounds.size.height - fillRect.size.height;
        
        fillPath = UIBezierPath(roundedRect: fillRect
            , byRoundingCorners: [UIRectCorner.bottomLeft, UIRectCorner.bottomRight]
            , cornerRadii: CGSize(width: 5, height: 5)).cgPath
        context.addPath(fillPath);
        context.setFillColor(fillColor.cgColor);
        context.fillPath();
    }
    
    func refresh() {
        self.lblBloodGroup?.text = self.record?.groupName
        self.lblFillPercentage?.text = "\(self.record?.fillPercentage ?? 0)%"
        self.setNeedsLayout();
    }
}
